modules/objdetect/test/test_qrcode_decode_encode.cpp

```diff
@@ -0,0 +1,251 @@
+                ASSERT_FALSE(decoded_info.empty());
+
+                std::string original_info = config["info"];
+                EXPECT_EQ(decoded_info, original_info);
+
+                return; // done
+            }
+        }
+        std::cerr
+            << "Not found results for '" << name_current_image
+            << "' image in config file:" << dataset_config << std::endl
+            << "Re-run tests with enabled UPDATE_DECODE_TEST_DATA macro to update test data."
+            << std::endl;
+    }
+}
+
+typedef testing::TestWithParam< std::string > Objdetect_QRCode_Encode;
+TEST_P(Objdetect_QRCode_Encode, regression){
+            const std::string name_current_image = GetParam();
+            const std::string root = "qrcode/decode_encode";
+
+            std::string image_path = findDataFile(root +"/"+ name_current_image);
+            /**read the original image */
+            Mat src = imread(image_path, IMREAD_GRAYSCALE), straight_barcode;
+            ASSERT_FALSE(src.empty()) << "Can't read image: " << image_path;
+            src = src(Range(border_width,src.rows-border_width),Range(border_width,src.rows-border_width)).clone();
+            resize(src,src,fixed_size);
+            const std::string dataset_config = findDataFile(root +"/"+ "dataset_config.json");
+            FileStorage file_config(dataset_config, FileStorage::READ);
+
+            ASSERT_TRUE(file_config.isOpened()) << "Can't read validation data: " << dataset_config;
+            {
+                FileNode images_list = file_config["test_images"];
+                size_t images_count = static_cast<size_t>(images_list.size());
+                ASSERT_GT(images_count, 0u) << "Can't find validation data entries in 'test_images': " << dataset_config;
+
+                for (size_t index = 0; index < images_count; index++)
+                {
+                    FileNode config = images_list[(int)index];
+                    std::string name_test_image = config["image_name"];
+                    if (name_test_image == name_current_image)
+                    {
+                        std::string original_info = config["info"];
+                        QRCodeEncoder encoder;
+                        Mat result = encoder.generateSingle(original_info,auto_mode);
+                        resize(result,result,fixed_size,0,0,INTER_AREA);
+                        std::vector<Point> corners;
+                        std::string decoded_info;
+
+                        for (int i = 0; i < 4; i++)
+                        {
+                            int x = config["x"][i];
+                            int y = config["y"][i];
+                            corners.push_back(Point(x,y));
+                        }
+                        /**the encoded info should be the same as the original info*/
+                        bool success = decodeQRCode(result, corners, decoded_info, straight_barcode);
```

Comment:
in this test you should do pixel-to-pixel comparison of generated and original images.
for encode and decode test (text-to-text) you should create another test with shuffling parameters, as we said earlier